sui_storage/src/lock_service.rs

```diff
@@ -0,0 +1,491 @@
+            let lock = lock.ok_or(SuiError::TransactionLockDoesNotExist)?;
+
+            if let Some(previous_tx_digest) = lock {
+                // Lock already set to different transaction
+                if previous_tx_digest != tx_digest {
+                    // TODO: add metrics here
+                    debug!(prev_tx_digest =? previous_tx_digest,
+                          cur_tx_digest =? tx_digest,
+                          "Conflicting transaction!  Lock state changed in unexpected way");
+                    return Err(SuiError::ConflictingTransaction {
+                        pending_transaction: previous_tx_digest,
+                    });
+                }
+            } else {
+                // Only write the locks that need to be written (are uninitialized)
+                let obj_ref = owned_input_objects[i];
+                locks_to_write.push((obj_ref, Some(tx_digest)));
+            }
+        }
+
+        if !locks_to_write.is_empty() {
+            self.transaction_lock
+                .batch()
+                .insert_batch(&self.transaction_lock, locks_to_write)?
+                .write()?;
+        }
+
+        Ok(())
+    }
+
+    /// Initialize a lock to None (but exists) for a given list of ObjectRefs.
+    /// If the lock already exists and is locked to a transaction, then return TransactionLockExists
+    fn initialize_locks(&self, objects: &[ObjectRef]) -> SuiResult {
+        // Use a multiget for efficiency
+        let locks = self.transaction_lock.multi_get(objects)?;
+
+        // If any locks exist and are not None, return errors for them
+        let existing_locks: Vec<ObjectRef> = locks
+            .iter()
+            .zip(objects)
+            .filter_map(|(lock_opt, objref)| {
+                if let Some(Some(_tx_digest)) = lock_opt {
+                    Some(*objref)
+                } else {
+                    None
+                }
+            })
+            .collect();
+        if !existing_locks.is_empty() {
+            return Err(SuiError::TransactionLockExists {
+                refs: existing_locks,
+            });
+        }
+
+        // Insert where locks don't exist already
+        let refs_to_insert = locks.iter().zip(objects).filter_map(|(lock_opt, objref)| {
+            if lock_opt.is_none() {
+                Some((objref, None))
+            } else {
+                None
+            }
+        });
+        self.transaction_lock.multi_insert(refs_to_insert)?;
+
+        Ok(())
+    }
+
+    /// Removes locks for a given list of ObjectRefs.
+    fn delete_locks(&self, objects: &[ObjectRef]) -> SuiResult {
+        self.transaction_lock.multi_remove(objects)?;
+        Ok(())
+    }
+
+    /// Loop to continuously process mutating commands in a single thread from async senders
+    fn command_loop(&self, receiver: Receiver<LockServiceCommands>) {
+        info!("LockService command processing loop started");
+        while let Ok(msg) = receiver.recv() {
```

Comment:
Lets add a comment here about this loop lifecycle. I think that it terminates when the sender closes / or when the sender closes and no more items are in the channel? Basically we need to make sure this task ends, when the program ends, and does so correctly (processes all data expected to process).

Reply:
I'm adding a comment.  The sender is not supposed to terminate until the end cuz this is a persistent service... or rather until the SuiDataStore is released, but it's up to the caller I suppose.  I don't know why they would terminate earlier.